src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsWriter.cs

```diff
@@ -0,0 +1,202 @@
+            await WriteBytesAsync(stream, _alloc.GetBuffer(len), 0, len, isAsync, ct).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Writes double value to out buffer, as little-endian.
+        /// </summary>
+        /// <param name="stream">Tds Stream to work with</param>
+        /// <param name="value">Value to write</param>
+        /// <param name="isAsync">Whether caller method is executing asynchronously.</param>
+        /// <param name="ct">Cancellation token</param>
+        public async ValueTask WriteDoubleAsync(TdsStream stream, double value, bool isAsync, CancellationToken ct)
+        {
+            Debug.Assert(!double.IsInfinity(value) && !double.IsNaN(value), "Double value is out of range.");
+
+            var len = sizeof(double);
+            BinaryPrimitives.WriteInt64LittleEndian(_alloc.GetBuffer(len), BitConverter.DoubleToInt64Bits(value));
+            await WriteBytesAsync(stream, _alloc.GetBuffer(len), 0, len, isAsync, ct).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Writes partial long value to out buffer, as little-endian. This should be ideally used for writing date, 
+        /// time and related data types. The length specified should be between 0 to 8 bytes supported by long.
+        /// </summary>
+        /// <param name="stream">Tds Stream to work with</param>
+        /// <param name="value">Value to write</param>
+        /// <param name="length">Length to fill</param>
+        /// <param name="isAsync">Whether caller method is executing asynchronously.</param>
+        /// <param name="ct">Cancellation token</param>
+        public async ValueTask WritePartialLongAsync(TdsStream stream, long value, int length, bool isAsync, CancellationToken ct)
+        {
+            Debug.Assert(length >= 0, "Length should not be negative");
+            Debug.Assert(length <= 8, "Length specified is longer than the size of a long");
+
+            var len = sizeof(long);
+            var buffer = _alloc.GetBuffer(len);
+            for (int i = 0; i < length; i++)
+            {
+                buffer[i] = (byte)((value >> (i * 8)) & 0xFF);
+            }
+            await WriteBytesAsync(stream, _alloc.GetBuffer(len), 0, length, isAsync, ct).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Writes bytes directly to _stream
+        /// </summary>
+        /// <param name="stream">Tds Stream to work with</param>
+        /// <param name="data">Bytes of data</param>
+        /// <param name="start">Offset of starting position</param>
+        /// <param name="length">Length of data</param>
+        /// <param name="isAsync">Whether caller method is executing asynchronously.</param>
+        /// <param name="ct">Cancellation token</param>
+        /// <returns></returns>
+        public async static ValueTask WriteBytesAsync(TdsStream stream, byte[] data, int start, int length, bool isAsync, CancellationToken ct)
```

Comment:
The `isAsync` flag forces us into this signature, but the signature means that we must always send byte arrays. This rules out the use of a simple `stackalloc`'d Span in an synchronous case, forcing allocations upstream.

I think it'd be better to have a pair of `WriteSpan` and `WriteMemoryAsync` methods in this case, pushing the sync/async decision upstream.

NB: there's some work in newer .NET which allows us to use `allows ref struct` in generic type constraints, which might have allowed a single `WriteBytesAsync<T>`, where `T` was either `Span<byte>` or `Memory<byte>`. This isn't permissible in .NET Framework, but might be worth considering in the far future.